File: classified/packages/plugin-plugin-manager/src/__tests__/e2e/test-utils.ts
import {
  type Content,
  type Entity,
  type IAgentRuntime,
  type Memory,
  type Room,
  asUUID,
  ChannelType,
  createUniqueUuid,
  EventType,
} from '@elizaos/core';
import { strict as assert } from 'node:assert';
import { v4 as uuid } from 'uuid';

/**
 * Sets up a standard scenario environment for plugin manager E2E tests.
 *
 * This function creates a user and a room, providing an
 * isolated environment for each test case.
 *
 * @param runtime The live IAgentRuntime instance provided by the TestRunner.
 * @returns A promise that resolves to an object containing the created user and room.
 */
export async function setupScenario(runtime: IAgentRuntime): Promise<{ user: Entity; room: Room }> {
  assert(runtime.agentId, 'Runtime must have an agentId to run a scenario');

  // 1. Create a test user entity
  const user: Entity = {
    id: asUUID(uuid()),
    names: ['Test User'],
    agentId: runtime.agentId,
    metadata: { type: 'user' },
  };
  await runtime.createEntity(user);
  assert(user.id, 'Created user must have an id');

  // 2. Create a test room
  const room: Room = {
    id: asUUID(uuid()),
    name: 'Test Plugin Manager Room',
    type: ChannelType.DM,
    source: 'e2e-test',
    serverId: 'e2e-test-server',
  };
  await runtime.createRoom(room);

  // 3. Ensure both the agent and the user are participants in the room
  await runtime.ensureParticipantInRoom(runtime.agentId, room.id);
  await runtime.ensureParticipantInRoom(user.id, room.id);

  return { user, room };
}

/**
 * Simulates a user sending a message and waits for the agent's response.
 *
 * This function abstracts the event-driven nature of the message handler
 * into a simple async function, making tests easier to write and read.
 *
 * @param runtime The live IAgentRuntime instance.
 * @param room The room where the message is sent.
 * @param user The user entity sending the message.
 * @param text The content of the message.
 * @returns A promise that resolves with the agent's response content.
 */
export function sendMessageAndWaitForResponse(
  runtime: IAgentRuntime,
  room: Room,
  user: Entity,
  text: string
): Promise<Content> {
  return new Promise((resolve) => {
    assert(runtime.agentId, 'Runtime must have an agentId to send a message');
    assert(user.id, 'User must have an id to send a message');

    // Construct the message object, simulating an incoming message from a user
    const message: Memory = {
      id: createUniqueUuid(runtime, `${user.id}-${Date.now()}`),
      agentId: runtime.agentId,
      entityId: user.id,
      roomId: room.id,
      content: {
        text,
      },
      createdAt: Date.now(),
    };

    // The callback function that the message handler will invoke with the agent's final response.
    // We use this callback to resolve our promise.
    const callback = (responseContent: Content) => {
      resolve(responseContent);
    };

    // Emit the event to trigger the agent's message processing logic.
    runtime.emitEvent(EventType.MESSAGE_RECEIVED, {
      runtime,
      message,
      callback,
    });
  });
}

File: cpp_merge/classified/packages/plugin-plugin-manager/src/__tests__/e2e/test-utils.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/plugin-plugin-manager/src/__tests__/e2e/test-utils.h"

std::shared_ptr<Promise<object>> setupScenario(std::shared_ptr<IAgentRuntime> runtime)
{
    assert(runtime->agentId, std:("Runtime must have an agentId to run a scenario"));
    auto user = object{
        object::pair{std:("id"), asUUID(uuid())}, 
        object::pair{std:("names"), array<string>{ std:("Test User") }}, 
        object::pair{std:("agentId"), runtime->agentId}, 
        object::pair{std:("metadata"), object{
            object::pair{std:("type"), std:("user")}
        }}
    };
    std::async([=]() { runtime->createEntity(user); });
    assert(user->id, std:("Created user must have an id"));
    auto room = object{
        object::pair{std:("id"), asUUID(uuid())}, 
        object::pair{std:("name"), std:("Test Plugin Manager Room")}, 
        object::pair{std:("type"), ChannelType->DM}, 
        object::pair{std:("source"), std:("e2e-test")}, 
        object::pair{std:("serverId"), std:("e2e-test-server")}
    };
    std::async([=]() { runtime->createRoom(room); });
    std::async([=]() { runtime->ensureParticipantInRoom(runtime->agentId, room->id); });
    std::async([=]() { runtime->ensureParticipantInRoom(user->id, room->id); });
    return object{
        object::pair{std:("user"), std:("user")}, 
        object::pair{std:("room"), std:("room")}
    };
};


std::shared_ptr<Promise<std::shared_ptr<Content>>> sendMessageAndWaitForResponse(std::shared_ptr<IAgentRuntime> runtime, std::shared_ptr<Room> room, std::shared_ptr<Entity> user, string text)
{
    return std::make_shared<Promise>([=](auto resolve) mutable
    {
        assert(runtime->agentId, std:("Runtime must have an agentId to send a message"));
        assert(user->id, std:("User must have an id to send a message"));
        auto message = object{
            object::pair{std:("id"), createUniqueUuid(runtime, string_empty + user->id + std:("-") + Date->now() + string_empty)}, 
            object::pair{std:("agentId"), runtime->agentId}, 
            object::pair{std:("entityId"), user->id}, 
            object::pair{std:("roomId"), room->id}, 
            object::pair{std:("content"), object{
                object::pair{std:("text"), std:("text")}
            }}, 
            object::pair{std:("createdAt"), Date->now()}
        };
        auto callback = [=](auto responseContent) mutable
        {
            resolve(responseContent);
        };
        runtime->emitEvent(EventType->MESSAGE_RECEIVED, object{
            object::pair{std:("runtime"), std:("runtime")}, 
            object::pair{std:("message"), std:("message")}, 
            object::pair{std:("callback"), std:("callback")}
        });
    }
    );
};


